transformer_engine/common/layer_norm/norms.h

```diff
@@ -0,0 +1,334 @@
+      return RMS_BWD_TUNED_FUNCS;
+  } else {
+    if constexpr (NormEnum == NVTE_NORM_TYPE::LN_FWD_TE)
+      return LN_FWD_GENERAL_FUNCS;
+    else if constexpr (NormEnum == NVTE_NORM_TYPE::LN_BWD_TE)
+      return LN_BWD_GENERAL_FUNCS;
+    else if constexpr (NormEnum == NVTE_NORM_TYPE::RMS_FWD_TE)
+      return RMS_FWD_GENERAL_FUNCS;
+    else if constexpr (NormEnum == NVTE_NORM_TYPE::RMS_BWD_TE)
+      return RMS_BWD_GENERAL_FUNCS;
+  }
+};
+
+template <typename T>
+struct TypeId {};
+
+template <>
+struct TypeId<fp16> {
+  constexpr static uint32_t Value = 0;
+};
+
+template <>
+struct TypeId<bf16> {
+  constexpr static uint32_t Value = 1;
+};
+
+template <>
+struct TypeId<fp32> {
+  constexpr static uint32_t Value = 2;
+};
+
+template <>
+struct TypeId<fp8e4m3> {
+  constexpr static uint32_t Value = 3;
+};
+
+template <typename T, int S>
+struct Type2Key {
+  constexpr static uint32_t Value = TypeId<T>::Value << S;
+};
+
+template <typename T>
+struct WeightType2Key : public Type2Key<T, 0> {};
+
+template <typename T>
+struct InputType2Key : public Type2Key<T, 2> {};
+
+template <typename T>
+struct OutputType2Key : public Type2Key<T, 4> {};
+
+template <typename T>
+struct ComputeType2Key : public Type2Key<T, 6> {};
+
+template <typename W, typename I, typename O, typename C>
+struct Types2Key {
+  constexpr static uint32_t Value = WeightType2Key<W>::Value | InputType2Key<I>::Value |
+                                    OutputType2Key<O>::Value | ComputeType2Key<C>::Value;
+  constexpr static inline uint64_t get(const uint64_t hidden_size) {
+    constexpr uint64_t type_key = Value;
+    return (type_key << 32) | hidden_size;
+  }
+};
+
+template <typename W, typename I, typename O, typename C, uint64_t HIDDEN_SIZE,
+          NVTE_NORM_TYPE NormEnum, bool IF_TUNED, typename Enable = void>
+struct NormRegistrar {};
+
+template <typename W, typename I, typename O, typename C, uint64_t HIDDEN_SIZE,
+          NVTE_NORM_TYPE NormEnum, bool IF_TUNED>
+struct NormRegistrar<W, I, O, C, HIDDEN_SIZE, NormEnum, IF_TUNED,
+                     typename std::enable_if<IF_TE_NORMS<NormEnum>()>::type> {
+  explicit NormRegistrar(typename LauncherType<NormEnum>::FunctionType f) {
+    auto& registry = GET_REGISTRY<NormEnum, IF_TUNED>();
+    if constexpr (IF_TUNED) {
+      uint64_t key = Types2Key<W, I, O, C>::get(HIDDEN_SIZE);
+      registry.insert({key, f});
+    } else {
+      uint64_t key = Types2Key<W, I, O, C>::get(0);
+      registry[key].insert({HIDDEN_SIZE, f});
+    }
+  }
+};
+
+class NormBase {
+ public:
+  virtual void initialize() = 0;
+
+  virtual void execute() = 0;
+
+  virtual void set_workspace_and_barrier(void* workspace_ptr, void* barrier_ptr) {};
+
+  virtual std::vector<size_t> get_workspace_shape() { return {0}; };
+
+  virtual std::vector<size_t> get_barrier_shape() { return {0}; };
+};
+
+template <NVTE_NORM_TYPE NormEnum>
+class NormFwdTe : public NormBase {
+ public:
+  NormFwdTe();
```

Comment:
I see that calling this constructor is considered an error:
https://github.com/NVIDIA/TransformerEngine/blob/41273ebec7941ddf12c785cd6f7113cf8c9062c0/transformer_engine/common/layer_norm/norms.cpp#L147-L152
We can catch this at compile-time with:
```suggestion
  NormFwdTe() = delete;
```